### PR TITLE
Constants file and AES-GCM key derivation function, version 2

### PR DESCRIPTION
**Pull request by matthewstevenson88**
This PR is identical to #4. I accidentally merged a bunch of commits into the other PR and had to close it. This PR has already been approved by @cesarghali and I still need approval from @beschmi.

The description from the previous PR is copied below:

This PR contains 2 distinct changes:

Build cleanup from the migration of the record protocol, and the addition of the s2a_constants.h file. Cesar can take a look at the constants file, but otherwise there is no review needed here.

The AES-GCM version of the HKDF_expand function from BoringSSL has been expanded to handle the more general cases that are needed for S2A. Note that the changes pass all the tests that are implemented (including the ALTS ones). Please both Cesar and Benedikt have a careful look at this part.

### CODE REVIEW

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 1]
```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
...
   HMAC_CTX_init(&hmac);
-  if (!HMAC_Init_ex(&hmac, kdf_key, kKdfKeyLen, EVP_sha256(), nullptr) ||
-      !HMAC_Update(&hmac, kdf_counter, kKdfCounterLen) ||
-      !HMAC_Update(&hmac, &ctr, 1) || !HMAC_Final(&hmac, buf, nullptr)) {
+  if (!HMAC_Init_ex(&hmac, prk, prk_size, digest, nullptr)) {
     HMAC_CTX_cleanup(&hmac);
     return GRPC_STATUS_INTERNAL;
   }
-  HMAC_CTX_cleanup(&hmac);
+  for (i = 0; i < n; i++) {
```
Comment (beschmi) [inline src/core/tsi/alts/crypt/aes_gcm.cc]: Can you create a function that takes an HMAC_CTX* and executes the loop (minus HMAC_CTX_cleanup/HMAC_CTX_free)? Then the two versions can call this function and clean up on their own.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 2]
```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
...
+    uint8_t ctr = i + 1;
```
Comment (beschmi) [inline src/core/tsi/alts/crypt/aes_gcm.cc]: Move down closer to first use.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 3]
```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
...
+                   !HMAC_Update(&hmac, buf, digest_size))) {
+      HMAC_CTX_cleanup(&hmac);
+      return GRPC_STATUS_INTERNAL;
+    }
+    if (!HMAC_Update(&hmac, info, info_size) || !HMAC_Update(&hmac, &ctr, 1) ||
+        !HMAC_Final(&hmac, buf, nullptr)) {
+      HMAC_CTX_cleanup(&hmac);
+      return GRPC_STATUS_INTERNAL;
+    }
+    todo = digest_size;
```
Comment (beschmi) [inline src/core/tsi/alts/crypt/aes_gcm.cc]: Declare todo here: size_t todo = ...;

File: src/core/tsi/s2a/s2a_constants.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+#ifndef GRPC_CORE_TSI_S2A_S2A_CONSTANTS_H
+#define GRPC_CORE_TSI_S2A_S2A_CONSTANTS_H
+
+/** The following constants are defined in BoringSSL but not OpenSSL.
+ *  See the aead.h file in BoringSSL for more documentation. **/
+#ifndef EVP_AEAD_AES_GCM_TAG_LEN
+#define EVP_AEAD_AES_GCM_TAG_LEN 16
+#endif
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/s2a/s2a_constants.h]: Done for all of `s2a_constants.h`.

File: src/core/tsi/alts/crypt/gsec.h [hunk 5]
```diff
@@ -451,6 +451,32 @@ grpc_status_code gsec_aes_gcm_aead_crypter_create(const uint8_t* key,
                                                   gsec_aead_crypter** crypter,
                                                   char** error_details);
 
+enum GsecHashFunction {
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/gsec.h]: Changed it. I found at least one instance of `enum class` in the C core: https://github.com/grpc/grpc/blob/master/src/core/lib/channel/channelz.h

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 6]
```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
...
 
-static grpc_status_code aes_gcm_derive_aead_key(uint8_t* dst,
-                                                const uint8_t* kdf_key,
-                                                const uint8_t* kdf_counter) {
-  unsigned char buf[EVP_MAX_MD_SIZE];
-  unsigned char ctr = 1;
+grpc_status_code hkdf_derive_secret(uint8_t* out_key, size_t out_size,
+                                     GsecHashFunction hash_function, const uint8_t* prk,
+                                     size_t prk_size, const uint8_t* info,
+                                     size_t info_size) {
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: We will do so, yes. The current plan is to not move ALTS files in the s2a-master fork, figure out what organizational changes we need to make, and then make the changes as a separate PR to gRPC master.

File: src/core/tsi/alts/crypt/aes_gcm.cc [hunk 7]
```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
...
+      break;
+    case SHA384_hash_function:
+      digest = EVP_sha384();
+      break;
+  }
+  const size_t digest_size = EVP_MD_size(digest);
+  uint8_t buf[EVP_MAX_MD_SIZE];
+  size_t n, done = 0;
+  unsigned i;
+  n = (out_size + digest_size - 1) / digest_size;
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: Done.

File: src/core/tsi/alts/crypt/aes_gcm.cc (hunk 3 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: Done.

File: src/core/tsi/alts/crypt/aes_gcm.cc (hunk 2 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: Done.

File: src/core/tsi/alts/crypt/aes_gcm.cc (hunk 1 above)
Comment (matthewstevenson88) [reply src/core/tsi/alts/crypt/aes_gcm.cc]: Done.

File: test/core/tsi/alts/crypt/aes_gcm_test.cc [hunk 8]
```diff
@@ -2093,13 +2093,94 @@ static void gsec_test_do_vector_tests_ieee() {
   gsec_aead_free_test_vector(test_vector_20);
 }
 
+static void hkdf_check_out_key(GsecHashFunction hash_function, uint8_t* info, size_t info_size,
+                               uint8_t* prk, size_t prk_size,
+                               uint8_t* correct_out_key, size_t out_key_size) {
+  uint8_t* out_key =
```
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: I'm not certain what you mean by this. Could you elaborate? I can set the test up similar to how `gsec_aead_test_vector` is used in the same test file, but that doesn't seem to be what you mean.

**Comment by matthewstevenson88** [comment]
Thanks for the comments @beschmi ! I've implemented most of them, and left follow up comments on the rest.

Also, please ignore the larger diffs - there were some changes to autogenerated files after I fetched from upstream.

File: src/core/tsi/s2a/s2a_constants.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (beschmi) [reply src/core/tsi/s2a/s2a_constants.h]: Can you follow the naming conventions from [1]? Then you also don't need the #ifndef anymore.

[1] https://google.github.io/styleguide/cppguide.html#Constant_Names

File: test/core/tsi/alts/crypt/aes_gcm_test.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (beschmi) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: I meant using std::vector for the test vector definition, function parameters and and local variables (like out_key). Then you can use vec.data() and vec.size() for the call to hkdf_derive_secret and you don't need to manually manage memory and you don't have to spell out the sizes for the test vectors.

File: src/core/tsi/s2a/s2a_constants.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply src/core/tsi/s2a/s2a_constants.h]: Done.

File: test/core/tsi/alts/crypt/aes_gcm_test.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (matthewstevenson88) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: In the newest commit, I've implemented my understanding of this. Please double check that it is what you intended. Thanks!

File: test/core/tsi/alts/crypt/aes_gcm_test.cc [hunk 9]
```diff
@@ -2093,13 +2094,85 @@ static void gsec_test_do_vector_tests_ieee() {
   gsec_aead_free_test_vector(test_vector_20);
 }
 
+static void hkdf_check_out_key(GsecHashFunction hash_function,
+                               std::vector<uint8_t>& info,
+                               std::vector<uint8_t>& prk,
+                               std::vector<uint8_t>& correct_out_key) {
+  std::vector<uint8_t> out_key;
```
Comment (beschmi) [inline test/core/tsi/alts/crypt/aes_gcm_test.cc]: Nit: Does ... out_key((correct_out_key.size(), 0); work?

File: test/core/tsi/alts/crypt/aes_gcm_test.cc [hunk 10]
```diff
@@ -2093,13 +2094,85 @@ static void gsec_test_do_vector_tests_ieee() {
...
+  out_key.resize(correct_out_key.size(), 0);
+  grpc_status_code derive_status =
+      hkdf_derive_secret(out_key.data(), out_key.size(), hash_function,
+                         prk.data(), prk.size(), info.data(), info.size());
+  GPR_ASSERT(derive_status == GRPC_STATUS_OK);
+  for (size_t i = 0; i < out_key.size(); i++) {
+    GPR_ASSERT(out_key[i] == correct_out_key[i]);
```
Comment (beschmi) [inline test/core/tsi/alts/crypt/aes_gcm_test.cc]: Can you use out_key == correct_out_key now?

Comment (matthewstevenson88) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: Done.

File: test/core/tsi/alts/crypt/aes_gcm_test.cc (hunk 9 above)
Comment (matthewstevenson88) [reply test/core/tsi/alts/crypt/aes_gcm_test.cc]: Done.

**Comment by matthewstevenson88** [comment]
Thanks for the review @beschmi! :)

**Pull request merged**